src/main/scala/viper/gobra/util/ViperChopper.scala

```diff
@@ -0,0 +1,644 @@
+  }
+
+  object Cut {
+
+    /**
+      * Returns a set of chopped programs where the number of programs is bounded by `bound`.
+      * @param N Number of nodes.
+      * @param nodes Nodes that must be included.
+      * @param edges Edges of the dependency graph.
+      * @param idToVertex Mapping from node ids to vertices used for price computations.
+      * @param bound Specifies upper bound on the number of returned programs.
+      *              If none, then maximum number of programs is returned.
+      * @param penalty Specifies penalty of merging programs.
+      * @return Set of programs.
+      */
+    def boundedCut[T](N: Int, nodes: Vector[Int], edges: Array[mutable.SortedSet[Int]], idToVertex: Int => Vertex)(
+      bound: Option[Int],
+      penalty: Penalty[Vertex]
+    ): Vector[List[Int]] = {
+      require(bound.forall(_ > 0), s"Got $bound as the size of the cut, but expected positive number")
+
+      val t1 = System.nanoTime()
+      val smallestPrograms = smallestCut(N, nodes, edges)
+      val t2 = System.nanoTime()
+      val mergedPrograms = mergePrograms(smallestPrograms)(bound, penalty.contravariantLift(idToVertex))
+      val t3 = System.nanoTime()
+      val timeCutting = BigDecimal((t2 - t1) / 1e9d).setScale(2, BigDecimal.RoundingMode.HALF_UP)
+      val timeMerging = BigDecimal((t3 - t2) / 1e9d).setScale(2, BigDecimal.RoundingMode.HALF_UP)
+
+      println(s"Chopped verification condition into ${mergedPrograms.size} parts. Maximum number of parts is ${smallestPrograms.size}. (Time: ${timeCutting}s + ${timeMerging}s)")
+
+      // Safety check validating the result partially. Can be removed in a year if no error has been found.
+      {
+        val containedInResult = Array.ofDim[Boolean](N)
+        for (program <- mergedPrograms; node <- program) { containedInResult(node) = true }
+        // check all nodes of the smallest programs are present (no node should be lost)
+        val smallestProgramContainedInResult = smallestPrograms.forall(_.forall(containedInResult(_)))
+        assert(smallestProgramContainedInResult, "Chopper Error: Lost nodes during merging step.")
+
+        val containedInSmallest = Array.ofDim[Boolean](N)
+        for (program <- smallestPrograms; node <- program) { containedInSmallest(node) = true }
+        // checks all selected notes are present in the result
+        val selectedNodesContainedInSmallest = nodes.forall(containedInSmallest(_))
+        assert(selectedNodesContainedInSmallest, "Chopper Error: Not all selected nodes present in solution.")
+      }
+
+      mergedPrograms
+    }
+
+    /**
+      * Returns the set of smallest possible programs.
+      * @param N Number of nodes.
+      * @param nodes Nodes that must be included.
+      * @param edges Edges of the dependency graph.
+      * @return Set of smallest possible programs.
+      */
+    def smallestCut(N: Int, nodes: Vector[Int], edges: Array[mutable.SortedSet[Int]]): Vector[List[Int]] = {
```

Comment:
same comment on ordered collection

Reply:
I do not know which collection you mean. The only ordered collection is the sorted set. Do you mean that I should use  the normal set instead of a sorted set? The reason why I use the sorted set is simply because set relies on hashes.

Reply:
I mean the return type. Why is it `Vector[List[Int]]` instead of e.g. `Set[List[Int]]`?

Reply:
Ah. I did not consider that you mean the Vector. Each of them is considered a separate program. With a set, it would check that all of them are different, which can be quite expensive and is unnecessary.

Reply:
Makes sense, I did not think about that getting the set properties is not free